buildpacks/ruby/src/layers/metrics_agent_install.rs

```diff
@@ -64,97 +60,53 @@ pub(crate) enum MetricsAgentInstallError {
     ChecksumFailed(String),
 }
 
-impl<'a> Layer for MetricsAgentInstall<'a> {
-    type Buildpack = RubyBuildpack;
-    type Metadata = Metadata;
-
-    fn types(&self) -> libcnb::data::layer_content_metadata::LayerTypes {
-        LayerTypes {
+pub(crate) fn install_metrics_agent(
+    context: &BuildContext<RubyBuildpack>,
+    mut bullet: Print<SubBullet<Stdout>>,
+) -> Result<(Print<SubBullet<Stdout>>, LayerEnv), libcnb::Error<RubyBuildpackError>> {
+    let new_metadata = Metadata {
+        download_url: Some(DOWNLOAD_URL.to_string()),
+    };
+
+    let layer = context.cached_layer(
+        layer_name!("metrics_agent"),
+        CachedLayerDefinition {
             build: true,
             launch: true,
-            cache: true,
+            invalid_metadata_action: &|_| InvalidMetadataAction::DeleteLayer,
+            restored_layer_action: &|old_metadata: &Metadata, _| {
+                if old_metadata == &new_metadata {
+                    RestoredLayerAction::KeepLayer
+                } else {
+                    RestoredLayerAction::DeleteLayer
+                }
+            },
+        },
+    )?;
+
+    let bin_dir = layer.path().join("bin");
+
+    match layer.state {
+        LayerState::Restored { .. } => {
+            bullet = bullet.sub_bullet("Using cached metrics agent");
         }
-    }
+        LayerState::Empty { .. } => {
```

Comment:
This doesn't handle the cases when the metrics agent is updated due to either another metrics agent URL, or invalid metadata